h2/src/main/org/h2/mvstore/tx/TransactionStore.java

```diff
@@ -245,25 +287,54 @@ public synchronized void close() {
      *
      * @return the transaction
      */
-    public synchronized Transaction begin() {
+    public Transaction begin() {
+        return begin(RollbackListener.NONE);
+    }
+
+    /**
+     * Begin a new transaction.
+     * @param listener to be notified in case of a rollback
+     *
+     * @return the transaction
+     */
+    public Transaction begin(RollbackListener listener) {
+        Transaction transaction = registerTransaction(0, Transaction.STATUS_OPEN, null, 0, listener);
+        return transaction;
+    }
+
+    private Transaction registerTransaction(int txId, int status, String name, long logId,
+                                            RollbackListener listener) {
 
         int transactionId;
-        int status;
-        if (!init) {
-            throw DataUtils.newIllegalStateException(
-                    DataUtils.ERROR_TRANSACTION_ILLEGAL_STATE,
-                    "Not initialized");
-        }
-        transactionId = openTransactions.nextClearBit(1);
-        if (transactionId > maxTransactionId) {
-            throw DataUtils.newIllegalStateException(
-                    DataUtils.ERROR_TOO_MANY_OPEN_TRANSACTIONS,
-                    "There are {0} open transactions",
-                    transactionId - 1);
-        }
-        openTransactions.set(transactionId);
-        status = Transaction.STATUS_OPEN;
-        return new Transaction(this, transactionId, status, null, 0);
+        long sequenceNo;
+        boolean success;
+        do {
+            VersionedBitSet original = openTransactions.get();
+            if (txId == 0) {
+                transactionId = original.nextClearBit(1);
+            } else {
+                transactionId = txId;
+                assert !original.get(transactionId);
+            }
+            if (transactionId > maxTransactionId) {
+                throw DataUtils.newIllegalStateException(
+                        DataUtils.ERROR_TOO_MANY_OPEN_TRANSACTIONS,
+                        "There are {0} open transactions",
+                        transactionId - 1);
+            }
+            VersionedBitSet clone = original.cloneIt();
+            clone.set(transactionId);
+            sequenceNo = clone.getVersion() + 1;
+            clone.setVersion(sequenceNo);
+            success = openTransactions.compareAndSet(original, clone);
+        } while(!success);
```

Comment:
I'm not sure that `success` variable here improves readability. Loops controlled by some flag usually take more time to understood where this flag is assigned even when assignment is nearby. `compareAndSet()` can be moved into `while` condition.

Reply:
Let me disagree about readability here: if variable is named approproately, it improves readability, at least for me. All of this is very subjective, of course. For instance, I start laughing when peope talk about improved readability with java8 streams, but some people love it.
The whole compareAndSet expression can’t be moved inside of the loop condition, because parameters would go out of scope, and early declaration of them makes it even more verbose and less readable.

Reply:
The comments in other places about `success` variable in loop conditions can be ignored too.